fluent-bundle/src/types.rs

```diff
@@ -111,14 +111,60 @@ impl<'source> From<&ast::InlineExpression<'source>> for DisplayableNode<'source>
     }
 }
 
-#[derive(Debug, PartialEq, Clone)]
+pub trait FluentType: std::fmt::Debug {
+    fn duplicate(&self) -> Box<dyn FluentType>;
+    fn get_hash(&self) -> (&'static str, usize);
+    fn to_string(&self) -> Cow<'static, str>;
+    fn display(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result;
+}
+
+impl PartialEq for dyn FluentType {
```

Comment:
A cleaner way to do this is

```rust
pub trait FluentType: AnyEq + 'static {
 // ...
}


impl PartialEq for dyn FluentType {
    fn eq(&self, other: &Self) -> bool {
        self.equals(other.as_any())
    }
}
pub trait AnyEq: Any + 'static {
    fn equals(&self, other: &dyn Any) -> bool;
    fn as_any(&self) -> &dyn Any;
}

impl<T: Any + PartialEq> AnyEq for T {
    fn equals(&self, other: &dyn Any) -> bool {
        if let Some(that) = other.downcast_ref::<Self>() {
            self == that
        } else {
            false
        }
    }
    fn as_any(&self) -> &dyn Any {
        self
    }
}

```
No need to have implementors write the hacky get_hash function.